### ISSUE TITLE
Implement new autowaits system 

### ISSUE THREAD

**Issue by inikulin** [DevExpress/testcafe#245]
We need better heuristics for page unloads: currently we always wait at least a second after the action. We've discussed this in person, @AlexanderMoskovkin can you write down proposed algorithm here, please?

**Comment by AlexanderMoskovkin** [comment]
There are some thoughts about this but there is no completed algorithm. I'd like to think out this more and write a proposed algorithm here a little later. The algorithm implementation depends on https://github.com/DevExpress/testcafe-hammerhead/issues/667 also

**Comment by AlexanderMoskovkin** [comment]
**Improve page unload barrier**

Now we wait for 400 ms after each action to be sure that action execution don't lead to page unloading.

We can improve this in the following way: we should wait this time only if there was one of the following page unload triggers:
 - Click on a link with src attribute
 - Form submit calling (or click on a submit button in the form)
 - Window/Document location changed

If there was no any of these things after an aciton execution the page unload barrier resolves waiting without any timeout.

_[optionally]_ As the second step of improvement we can check if setTimeout/setInterval functions were called on action handlers. If they were called we prolong waiting for page unload triggers (we assume that this triggers can be called after timeout)

**Comment by inikulin** [comment]
@AlexanderMoskovkin Can we also catch canceled events? @churkin said that it's quite complicated, but as far as I understand we can still do it in some cases? It would be nice to have that heuristics as well.

**Comment by AlexanderMoskovkin** [comment]
There are a lot of scenarios, I think we can cover most commons. We'll investigate them in more detail when we are starting to implement this.

**Comment by AlexanderMoskovkin** [comment]
blocked on https://github.com/DevExpress/testcafe-hammerhead/issues/879
